File: src/main/java/com/dh/ondot/schedule/api/response/AlarmDto.java
package com.dh.ondot.schedule.api.response;

import com.dh.ondot.core.util.TimeUtils;
import com.dh.ondot.schedule.domain.Alarm;

import java.time.LocalDateTime;

public record AlarmDto(
        Long alarmId,
        String alarmMode,
        boolean isEnabled,
        LocalDateTime triggeredAt,
        boolean isSnoozeEnabled,
        Integer snoozeInterval,
        Integer snoozeCount,
        String soundCategory,
        String ringTone,
        Double volume
) {
    public static AlarmDto of(Alarm alarm) {
        return new AlarmDto(
                alarm.getId(),
                alarm.getMode().name(),
                alarm.isEnabled(),
                TimeUtils.toSeoulDateTime(alarm.getTriggeredAt()),
                alarm.getSnooze().isSnoozeEnabled(),
                alarm.getSnooze().getSnoozeInterval().getValue(),
                alarm.getSnooze().getSnoozeCount().getValue(),
                alarm.getSound().getSoundCategory().name(),
                alarm.getSound().getRingTone().name(),
                alarm.getSound().getVolume()
        );
    }
}

File: src/main/kotlin/com/dh/ondot/schedule/presentation/response/AlarmDto.kt
package com.dh.ondot.schedule.presentation.response

import com.dh.ondot.core.util.TimeUtils
import com.dh.ondot.schedule.domain.Alarm
import com.dh.ondot.schedule.domain.Schedule
import java.time.LocalDateTime

data class AlarmDto(
    val alarmId: Long?,
    val alarmMode: String,
    val isEnabled: Boolean,
    val triggeredAt: LocalDateTime?,
    val isSnoozeEnabled: Boolean,
    val snoozeInterval: Int,
    val snoozeCount: Int,
    val soundCategory: String,
    val ringTone: String,
    val volume: Double,
) {
    companion object {
        @JvmStatic
        fun of(alarm: Alarm): AlarmDto {
            return AlarmDto(
                alarmId = alarm.id,
                alarmMode = alarm.mode.name,
                isEnabled = alarm.isEnabled,
                triggeredAt = TimeUtils.toSeoulDateTime(alarm.triggeredAt),
                isSnoozeEnabled = alarm.snooze.isSnoozeEnabled,
                snoozeInterval = alarm.snooze.snoozeInterval.value,
                snoozeCount = alarm.snooze.snoozeCount.value,
                soundCategory = alarm.sound.soundCategory.name,
                ringTone = alarm.sound.ringTone.name,
                volume = alarm.sound.volume,
            )
        }

        @JvmStatic
        fun of(alarm: Alarm, schedule: Schedule): AlarmDto {
            return AlarmDto(
                alarmId = alarm.id,
                alarmMode = alarm.mode.name,
                isEnabled = alarm.isEnabled,
                triggeredAt = TimeUtils.toSeoulDateTime(schedule.getNextRepeatAlarmTime(alarm.triggeredAt)),
                isSnoozeEnabled = alarm.snooze.isSnoozeEnabled,
                snoozeInterval = alarm.snooze.snoozeInterval.value,
                snoozeCount = alarm.snooze.snoozeCount.value,
                soundCategory = alarm.sound.soundCategory.name,
                ringTone = alarm.sound.ringTone.name,
                volume = alarm.sound.volume,
            )
        }
    }
}
